core/state/codestore.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2023 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package state
+
+import (
+	"errors"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/lru"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/ethdb"
+)
+
+const (
+	// Number of codehash->size associations to keep.
+	codeSizeCacheSize = 100000
+
+	// Cache size granted for caching clean code.
+	codeCacheSize = 64 * 1024 * 1024
+)
+
+var errMismatchedLength = errors.New("provided lists have different lengths")
+
+// CodeDB is an implementation of the CodeStore interface, designed for providing
+// efficient read and write functionalities for contract code.
+type CodeDB struct {
+	db    ethdb.KeyValueStore
+	size  *lru.Cache[common.Hash, int]
+	cache *lru.SizeConstrainedCache[common.Hash, []byte]
+}
+
+// NewCodeDB returns a codeDB instance with given database.
+func NewCodeDB(db ethdb.KeyValueStore) *CodeDB {
+	return &CodeDB{
+		db:    db,
+		size:  lru.NewCache[common.Hash, int](codeSizeCacheSize),
+		cache: lru.NewSizeConstrainedCache[common.Hash, []byte](codeCacheSize),
+	}
+}
+
+// ReadCode implements CodeReader, retrieving a particular contract's code
+// with given contract address and code hash.
+func (db *CodeDB) ReadCode(address common.Address, codeHash common.Hash) ([]byte, error) {
```

Comment:
I'd like to have an `offset` and `length` parameter here so that it can abstract whether the underlying code is chunked or not.

Reply:
I will try to make something and let you confirm it's what you want.

Reply:
Btw, I did some quick prototype for this idea locally and gave up finally. I don't understand why do we want to read a slice of contract code for EVM execution.

For a full-node, the whole contract code is present, we can just load the whole code in a single read. For a stateless client, it only has a batch of code slices with the associated code positions(e.g. offset, length, etc).  The assumption is that for this EVM execution, only this part of contract code will be accessed, and out-of-slice error should be raised if not.

So I am wondering if we should go in another direction. The code returned by the `CodeStore` interface should always be "complete". The one returned by stateless client can be padded with zero bytes along with the code position information. In another word, we need a `Code` interface to hide the difference between full code and partial code.

Reply:
yes padding is what we intend to do with the status quo. But this makes the verkle code much more inefficient because there will be a lot of copies each time. What I think should happen, is that we call `ReadCode` with an offset each time we need to read a chunk of data. For `EXTCODECOPY` we would always have this inefficency problem, but most of the time, we would not because it's mostly the interpreter loop.

Reply:
> because there will be a lot of copies each time

Copy contract code(I guess it's up to 2 Megabytes?) should be OKish, at least it's same as we have now?

> What I think should happen, is that we call ReadCode with an offset each time we need to read a chunk of data

Read the specific code chunk we need for vm interpreting is ideal, my concern is it might result in a big change in vm implementation. Although I am not confident to say it loudly :)

Just throw out my 2cents, I am totally fine to introduce code range notion for sure!